Flaky test: test_storage_kafka/test_batch_slow_6.py::test_kafka_rebalance[generate_new_create_table_query-{}.*Saved offset]

_Important: This issue was automatically generated and is used by CI for matching failures. DO NOT modify the body content. DO NOT remove labels._

Test name: test_storage_kafka/test_batch_slow_6.py::test_kafka_rebalance[generate_new_create_table_query-{}.*Saved offset]
Failure reason: ThreadSanitizer: data race contrib/librdkafka/src/rdkafka_queue.h
CI report: [Integration tests (amd_tsan, 2/6)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=102342&sha=5c81fd3baa44cee420e3953dac1cfc927e4161f6&name_0=PR&name_1=Integration%20tests%20%28amd_tsan%2C%202%2F6%29)

Failing test history: [cidb](https://play.clickhouse.com#V0lUSAogICAgOTAgQVMgaW50ZXJ2YWxfZGF5cwpTRUxFQ1QKICAgIHRvU3RhcnRPZkRheShjaGVja19zdGFydF90aW1lKSBBUyBkYXksCiAgICBjb3VudCgpIEFTIGZhaWx1cmVzLAogICAgZ3JvdXBVbmlxQXJyYXkocHVsbF9yZXF1ZXN0X251bWJlcikgQVMgcHJzLAogICAgYW55KHJlcG9ydF91cmwpIEFTIHJlcG9ydF91cmwKRlJPTSBjaGVja3MKV0hFUkUgKG5vdygpIC0gdG9JbnRlcnZhbERheShpbnRlcnZhbF9kYXlzKSkgPD0gY2hlY2tfc3RhcnRfdGltZQogICAgQU5EIHRlc3RfbmFtZSA9ICd0ZXN0X3N0b3JhZ2Vfa2Fma2EvdGVzdF9iYXRjaF9zbG93XzYucHk6OnRlc3Rfa2Fma2FfcmViYWxhbmNlW2dlbmVyYXRlX25ld19jcmVhdGVfdGFibGVfcXVlcnkte30uKlNhdmVkIG9mZnNldF0nCiAgICAtLSBBTkQgY2hlY2tfbmFtZSA9ICdJbnRlZ3JhdGlvbiB0ZXN0cyAoYW1kX3RzYW4sIDIvNiknCiAgICBBTkQgdGVzdF9zdGF0dXMgSU4gKCdGQUlMJywgJ0VSUk9SJykKICAgIEFORCAocHVsbF9yZXF1ZXN0X251bWJlciA9IDAgT1IgYmFzZV9yZWYgSU4gKCdtYXN0ZXInKSkKICAgIEFORCB0ZXN0X2NvbnRleHRfcmF3IExJS0UgJyVUaHJlYWRTYW5pdGl6ZXI6IGRhdGEgcmFjZSBjb250cmliL2xpYnJka2Fma2Evc3JjL3Jka2Fma2FfcXVldWUuaCUnCkdST1VQIEJZIGRheQpPUkRFUiBCWSBkYXkgREVTQwo=)

Test output:
```
File: test_storage_kafka/test_batch_slow_6.py:35 - in kafka_cluster
    cluster.shutdown()
File: helpers/cluster.py:4206 - in shutdown
    raise Exception(
E   Exception: Sanitizer assert found for instance ==================
E   WARNING: ThreadSanitizer: data race (pid=8)
E     Write of size 4 at 0x722c00237384 by thread T503 (mutexes: write M0, write M1):
E       #0 rd_kafka_q_yield contrib/librdkafka/src/rdkafka_queue.h:364:32 (clickhouse+0x304330a7) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #1 rd_kafka_timer_schedule_next contrib/librdkafka/src/rdkafka_timer.c:87:25 (clickhouse+0x30432379) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #2 rd_kafka_timer_schedule contrib/librdkafka/src/rdkafka_timer.c:113:9 (clickhouse+0x30432379)
E       #3 rd_kafka_timer_start0 contrib/librdkafka/src/rdkafka_timer.c:195:9 (clickhouse+0x304321ea) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #4 rd_kafka_metadata_fast_leader_query contrib/librdkafka/src/rdkafka_metadata.c:1733:17 (clickhouse+0x3034265e) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #5 rd_kafka_toppar_leader_unavailable contrib/librdkafka/src/rdkafka_partition.c:2585:9 (clickhouse+0x303c5724) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #6 rd_kafka_fetch_reply_handle_partition_error contrib/librdkafka/src/rdkafka_fetcher.c:255:17 (clickhouse+0x3032ba78) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #7 rd_kafka_fetch_reply_handle_partition contrib/librdkafka/src/rdkafka_fetcher.c:712:17 (clickhouse+0x30326964) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #8 rd_kafka_fetch_reply_handle contrib/librdkafka/src/rdkafka_fetcher.c:839:29 (clickhouse+0x30320faa) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #9 rd_kafka_broker_fetch_reply contrib/librdkafka/src/rdkafka_fetcher.c:909:23 (clickhouse+0x3031f132) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #10 rd_kafka_buf_callback contrib/librdkafka/src/rdkafka_buf.c:515:17 (clickhouse+0x302d64d2) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #11 rd_kafka_req_response contrib/librdkafka/src/rdkafka_broker.c:1974:9 (clickhouse+0x302bea7e) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #12 rd_kafka_recv contrib/librdkafka/src/rdkafka_broker.c:2093:17 (clickhouse+0x302bea7e)
E       #13 rd_kafka_transport_io_event contrib/librdkafka/src/rdkafka_transport.c:748:32 (clickhouse+0x3043ae37) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #14 rd_kafka_transport_io_serve contrib/librdkafka/src/rdkafka_transport.c:1007:17 (clickhouse+0x3043ae37)
E       #15 rd_kafka_broker_ops_io_serve contrib/librdkafka/src/rdkafka_broker.c:3670:21 (clickhouse+0x302ce7d4) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #16 rd_kafka_broker_consumer_serve contrib/librdkafka/src/rdkafka_broker.c:4373:21 (clickhouse+0x302cb55d) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #17 rd_kafka_broker_serve contrib/librdkafka/src/rdkafka_broker.c:4515:17 (clickhouse+0x302cb55d)
E       #18 rd_kafka_broker_thread_main contrib/librdkafka/src/rdkafka_broker.c:4715:25 (clickhouse+0x302c45e7) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #19 _thrd_wrapper_function contrib/librdkafka/src/tinycthread.c:576:9 (clickhouse+0x3044c63b) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E   
E     Previous read of size 4 at 0x722c00237384 by thread T501:
E       #0 rd_kafka_q_serve contrib/librdkafka/src/rdkafka_queue.c:524:13 (clickhouse+0x303cd558) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #1 rd_kafka_thread_main contrib/librdkafka/src/rdkafka.c:2155:17 (clickhouse+0x302dc0b0) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #2 _thrd_wrapper_function contrib/librdkafka/src/tinycthread.c:576:9 (clickhouse+0x3044c63b) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E   
E     Location is heap block of size 176 at 0x722c00237300 allocated by thread T149:
E       #0 malloc /ClickHouse/contrib/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:670:5 (clickhouse+0xbe75580) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #1 rd_malloc contrib/librdkafka/src/rd.h:142:19 (clickhouse+0x303cb82e) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #2 rd_kafka_q_new0 contrib/librdkafka/src/rdkafka_queue.c:117:29 (clickhouse+0x303cb82e)
E       #3 rd_kafka_new contrib/librdkafka/src/rdkafka.c:2295:34 (clickhouse+0x302d980f) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #4 cppkafka::Consumer::Consumer(cppkafka::Configuration) contrib/cppkafka/src/consumer.cpp:65:23 (clickhouse+0x302b10c1) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #5 cppkafka::Consumer* std::__1::construct_at[abi:fqe220101]<cppkafka::Consumer, cppkafka::Configuration, cppkafka::Consumer*>(cppkafka::Consumer*, cppkafka::Configuration&&) contrib/llvm-project/libcxx/include/__memory/construct_at.h:37:49 (clickhouse+0x1ead6237) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #6 cppkafka::Consumer* std::__1::__construct_at[abi:fqe220101]<cppkafka::Consumer, cppkafka::Configuration, cppkafka::Consumer*>(cppkafka::Consumer*, cppkafka::Configuration&&) contrib/llvm-project/libcxx/include/__memory/construct_at.h:45:10 (clickhouse+0x1ead6237)
E       #7 void std::__1::allocator_traits<std::__1::allocator<cppkafka::Consumer>>::construct[abi:fqe220101]<cppkafka::Consumer, cppkafka::Configuration, 0>(std::__1::allocator<cppkafka::Consumer>&, cppkafka::Consumer*, cppkafka::Configuration&&) contrib/llvm-project/libcxx/include/__memory/allocator_traits.h:302:5 (clickhouse+0x1ead6237)
E       #8 std::__1::__shared_ptr_emplace<cppkafka::Consumer, std::__1::allocator<cppkafka::Consumer>>::__shared_ptr_emplace[abi:fqe220101]<cppkafka::Configuration, std::__1::allocator<cppkafka::Consumer>, 0>(std::__1::allocator<cppkafka::Consumer>, cppkafka::Configuration&&) contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:161:5 (clickhouse+0x1ead6237)
E       #9 std::__1::shared_ptr<cppkafka::Consumer> std::__1::allocate_shared[abi:fqe220101]<cppkafka::Consumer, std::__1::allocator<cppkafka::Consumer>, cppkafka::Configuration, 0>(std::__1::allocator<cppkafka::Consumer> const&, cppkafka::Configuration&&) contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:682:51 (clickhouse+0x1ead6237)
E       #10 std::__1::shared_ptr<cppkafka::Consumer> std::__1::make_shared[abi:fqe220101]<cppkafka::Consumer, cppkafka::Configuration, 0>(cppkafka::Configuration&&) contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:690:10 (clickhouse+0x1eafb161) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #11 DB::KafkaConsumer2::createConsumer(cppkafka::Configuration) src/Storages/Kafka/KafkaConsumer2.cpp:90:16 (clickhouse+0x1eafb161)
E       #12 DB::KeeperHandlingConsumer::startUsing(std::__1::function<cppkafka::Configuration (std::__1::shared_ptr<DB::IKafkaExceptionInfoSink>)> const&) src/Storages/Kafka/KeeperHandlingConsumer.cpp:509:25 (clickhouse+0x1eaefb36) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #13 DB::StorageKafka2::acquireConsumer(unsigned long) src/Storages/Kafka/StorageKafka2.cpp:1142:45 (clickhouse+0x1ebc2120) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #14 DB::StorageKafka2::streamToViews(unsigned long) src/Storages/Kafka/StorageKafka2.cpp:1092:21 (clickhouse+0x1ebbf69d) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
E       #15 DB::StorageKafka2::threadFunc(unsigned long) src/Storages/Kafka/StorageKafka2.cpp:1041:42 (clickhouse+0x1ebbd946) (BuildId: 500bce84bd1ddc86794d7d99e5d8e45995bf9bfb)
~~~~~ truncated 234 lines ~~~~~
```

--- a/src/Storages/Kafka/KafkaConsumer2.cpp
+++ b/src/Storages/Kafka/KafkaConsumer2.cpp
@@ -149,14 +149,24 @@ void KafkaConsumer2::cleanQueuesAndMessages()
 void KafkaConsumer2::initializeQueues(const cppkafka::TopicPartitionList & topic_partitions)
 {
     cleanQueuesAndMessages();
-    // cppkafka itself calls assign(), but in order to detach the queues here we have to do the assignment manually.
-    // Later on we have to reassign the topic partitions with correct offsets.
-    consumer->assign(topic_partitions);
-    for (const auto & topic_partition : topic_partitions)
-        // This will also detach the partition queues from the consumer, thus the messages won't be forwarded without
-        // attaching them manually
-        queues.emplace(
-            TopicPartition{topic_partition.get_topic(), topic_partition.get_partition()}, consumer->get_partition_queue(topic_partition));
+
+    queues.reserve(topic_partitions.size());
+    // `get_partition_queue` creates the partition queue if needed and disables forwarding to the main consumer queue.
+    // Do this before `assign` starts fetching, otherwise messages may leak into the main queue before we detach it.
+    try
+    {
+        for (const auto & topic_partition : topic_partitions)
+            queues.emplace(
+                TopicPartition{topic_partition.get_topic(), topic_partition.get_partition()},
+                consumer->get_partition_queue(topic_partition));
+
+        consumer->assign(topic_partitions);
+    }
+    catch (...)
+    {
+        cleanQueuesAndMessages();
+        throw;
+    }
 }
 
 // it does the poll when needed

